<a href="https://colab.research.google.com/github/SelvamRaju/Natural-Language-Processing-Statistical-NLP/blob/main/Statistical_NLP_Part_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Domain : Digital Content Management

In [1]:
# Imports
import pandas as pd
import numpy as np
import os

import warnings
warnings.filterwarnings("ignore")

1 . Read and Analyse Dataset.

  A. Clearly write outcome of data analysis(Minimum 2 points)

  B. Clean the Structured Data

    i. Missing value analysis and imputation.

    ii. Eliminate Non-English textual data.
    Hint: Refer ‘langdetect’ library to detect language of the input text)

In [2]:
# Drive mounting
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [3]:
#Declaring the working directory (Directory in Drive)
os.chdir('/content/drive/MyDrive/Colab Notebooks/Data/Statistical-NLP/');

In [4]:
#Listing the files under the current working directory
!ls

blogtext.csv


In [5]:
#Importing CSV dataset
df = pd.read_csv('blogtext.csv')
df.head()

id  ...                                               text
0  2059027  ...             Info has been found (+/- 100 pages,...
1  2059027  ...             These are the team members:   Drewe...
2  2059027  ...             In het kader van kernfusie op aarde...
3  2059027  ...                   testing!!!  testing!!!          
4  3581210  ...               Thanks to Yahoo!'s Toolbar I can ...

[5 rows x 7 columns]

In [6]:
# Shape of dataset
df.shape

(681284, 7)

In [7]:
# Getting unique topics
df['topic'].unique()

array(['Student', 'InvestmentBanking', 'indUnk', 'Non-Profit', 'Banking',
       'Education', 'Engineering', 'Science', 'Communications-Media',
       'BusinessServices', 'Sports-Recreation', 'Arts', 'Internet',
       'Museums-Libraries', 'Accounting', 'Technology', 'Law',
       'Consulting', 'Automotive', 'Religion', 'Fashion', 'Publishing',
       'Marketing', 'LawEnforcement-Security', 'HumanResources',
       'Telecommunications', 'Military', 'Government', 'Transportation',
       'Architecture', 'Advertising', 'Agriculture', 'Biotech',
       'RealEstate', 'Manufacturing', 'Construction', 'Chemicals',
       'Maritime', 'Tourism', 'Environment'], dtype=object)

In [8]:
# Count of unique topics
len(df['topic'].unique())

40

In [9]:
# Count from groupby topic
df.groupby(['topic']).count()

id  gender     age    sign    date    text
topic                                                                  
Accounting                 3832    3832    3832    3832    3832    3832
Advertising                4676    4676    4676    4676    4676    4676
Agriculture                1235    1235    1235    1235    1235    1235
Architecture               1638    1638    1638    1638    1638    1638
Arts                      32449   32449   32449   32449   32449   32449
Automotive                 1244    1244    1244    1244    1244    1244
Banking                    4049    4049    4049    4049    4049    4049
Biotech                    2234    2234    2234    2234    2234    2234
BusinessServices           4500    4500    4500    4500    4500    4500
Chemicals                  3928    3928    3928    3928    3928    3928
Communications-Media      20140   20140   20140   20140   20140   20140
Construction               1093    1093    1093    1093    1093    1093
Consulting                 5862    5862    5862    5862    5862    5862
Education                 29633   29633   29633   29633   29633   29633
Engineering               11653   11653   11653   11653   11653   11653
Environment                 592     592     592     592     592     592
Fashion                    4851    4851    4851    4851    4851    4851
Government                 6907    6907    6907    6907    6907    6907
HumanResources             3010    3010    3010    3010    3010    3010
Internet                  16006   16006   16006   16006   16006   16006
InvestmentBanking          1292    1292    1292    1292    1292    1292
Law                        9040    9040    9040    9040    9040    9040
LawEnforcement-Security    1878    1878    1878    1878    1878    1878
Manufacturing              2272    2272    2272    2272    2272    2272
Maritime                    280     280     280     280     280     280
Marketing                  4769    4769    4769    4769    4769    4769
Military                   3128    3128    3128    3128    3128    3128
Museums-Libraries          3096    3096    3096    3096    3096    3096
Non-Profit                14700   14700   14700   14700   14700   14700
Publishing                 7753    7753    7753    7753    7753    7753
RealEstate                 2870    2870    2870    2870    2870    2870
Religion                   5235    5235    5235    5235    5235    5235
Science                    7269    7269    7269    7269    7269    7269
Sports-Recreation          3038    3038    3038    3038    3038    3038
Student                  153903  153903  153903  153903  153903  153903
Technology                42055   42055   42055   42055   42055   42055
Telecommunications         3891    3891    3891    3891    3891    3891
Tourism                    1942    1942    1942    1942    1942    1942
Transportation             2326    2326    2326    2326    2326    2326
indUnk                   251015  251015  251015  251015  251015  251015

In [10]:
#info about dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 681284 entries, 0 to 681283
Data columns (total 7 columns):
 #   Column  Non-Null Count   Dtype 
---  ------  --------------   ----- 
 0   id      681284 non-null  int64 
 1   gender  681284 non-null  object
 2   age     681284 non-null  int64 
 3   topic   681284 non-null  object
 4   sign    681284 non-null  object
 5   date    681284 non-null  object
 6   text    681284 non-null  object
dtypes: int64(2), object(5)
memory usage: 36.4+ MB


In [11]:
#null check
df.isnull().any()

id        False
gender    False
age       False
topic     False
sign      False
date      False
text      False
dtype: bool

In [12]:
# Displaying the datatypes
df.dtypes.to_frame(name = 'Data Type')

Data Type
id         int64
gender    object
age        int64
topic     object
sign      object
date      object
text      object

**Data Analysis outcome:**


*   *The dataset has 681284 rows of data with 7 variable/attributes of information*
*   *There are 40 unique topics where all the ~681K row fall under.*
*   *id and age are the 2 numerice variables*


No missing values identified in the analysis

In [16]:
# Cropping the dataset for computation. Taking the 5K into account for trainning

cropped_df = df[:5000]
cropped_df.shape

(5000, 7)

In [14]:
# Eliminating non english textual data using langdetect library

!pip install langdetect
from langdetect import detect

def detect_english(text):
    try:
        return detect(text) == 'en'
    except:
        return False

In [17]:
cropped_df = cropped_df[cropped_df['text'].apply(detect_english)]
cropped_df.shape

(4721, 7)

**Langdetect has eliminated some of the non-english textual data and out of 5000 we now have 4721 english data for computation.**

2 . Preprocess unstructured data to make it consumable for model training.

  A. Eliminate All special Characters and Numbers

  B. Lowercase all textual data

  C. Remove all Stopwords

  D. Remove all extra white spaces


In [62]:
# Eliminating stop words using NLTK and removing white spaces and converting texts to lower case

import nltk
nltk.download('stopwords')
from nltk.corpus import stopwords
import string, re

stop_words = set(stopwords.words('english'))

# Printing all stop words
print('Number of stop words : ', len(stop_words))

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
Number of stop words :  179


In [63]:
def text_cleaner(string):
  try:
   string=re.sub(r'^https?:\/\/<>.*[\r\n]*','',string,flags=re.MULTILINE) #Removing Special characters
   words=string.strip().lower().split() # lowercase conversion
   words=[w for w in words if not w in stop_words] #Removing stopwords
   return " ".join(words) # removes space 
  except:
    return ""

In [64]:
cropped_df['processed_text'] = cropped_df['text'].map(text_cleaner) 

In [65]:
# Displaying Text Comparison- using tail to view last rows

cropped_df[['processed_text','text']].tail(2)

processed_text                                               text
4998  ok lied fed playing friday night underdog cons...         ok  so i lied    fed up isn t playing f...
4999  well today went church talked music director t...         well  today i went to church and talked...

3 . Build a base Classification model

  A. Create dependent and independent variables
    
    Hint: Treat ‘topic’ as a Target variable.

  B. Split data into train and test.

  C. Vectorize data using any one vectorizer.

  D. Build a base model for Supervised Learning - Classification.

  E. Clearly print Performance Metrics.
    
    Hint: Accuracy, Precision, Recall, ROC-AUC

In [66]:
# Merging gender, age,topic and sign variables to make a grouped label for multilabel classification
cropped_df['labels'] = cropped_df['gender'] +',' + cropped_df['age'].astype(str) + ',' + cropped_df['topic'] + ',' + cropped_df['sign']

cropped_df['labels'].head()

0                   male,15,Student,Leo
2                   male,15,Student,Leo
3                   male,15,Student,Leo
4    male,33,InvestmentBanking,Aquarius
5    male,33,InvestmentBanking,Aquarius
Name: labels, dtype: object

In [70]:
processing_df = cropped_df[['processed_text', 'labels']]
processing_df.head()

processed_text                              labels
0  info found pages mb pdf files wait untill team...                 male,15,Student,Leo
2  het kader van kernfusie op aarde maak je eigen...                 male,15,Student,Leo
3                                    testing testing                 male,15,Student,Leo
4  thanks yahoo toolbar capture urls popups means...  male,33,InvestmentBanking,Aquarius
5  interesting conversation dad morning talking k...  male,33,InvestmentBanking,Aquarius

In [75]:
# Train Test split
from sklearn.model_selection import train_test_split

X = processing_df['processed_text']
y = processing_df['labels']

X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.3, random_state=3)


In [77]:
# Vectorizing using Count Vectorizer
from sklearn.feature_extraction.text import CountVectorizer
count_vectorizer = CountVectorizer(ngram_range=(1, 2),min_df=2)
count_vectorizer.fit(X_train)

CountVectorizer(min_df=2, ngram_range=(1, 2))

In [78]:
# Printing the length of vocabulary
len(count_vectorizer.vocabulary_)

32357

4 . Improve Performance of model.

  A. Experiment with other vectorisers.

  B. Build classifier Models using other algorithms than base model.

  C. Tune Parameters/Hyperparameters of the model/s.

  D. Clearly print Performance Metrics.

  Hint: Accuracy, Precision, Recall, ROC-AUC

5 . Share insights on relative performance comparison

  A. Which vectorizer performed better? Probable reason?.

  B. Which model outperformed? Probable reason?

  C. Which parameter/hyperparameter significantly helped to improve performance?Probable reason?.

  D. According to you, which performance metric should be
given most importance, why?.